File: sorting/java/QuickSort.java
/*
   Course: CS 41600
   Professor: Dr. Dai
   Name: Jackson Trader
   ID: 17
   Date: 2/23/2026
   File: QuickSort.java

   Description:
      This program implements the Quick Sort algorithm to sort a list of
      integers. It reads integers from standard input (one per line), sorts
      them in ascending order using Quick Sort with a last-element pivot
      and Lomuto partition scheme, and prints the sorted result to standard
      output (one integer per line). Input/output redirection is used to
      read from and write to files.
*/

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.io.IOException;
import java.util.ArrayList;

public class QuickSort {
    /**
     * Reads integers from standard input, one per line.
     * 
     * Uses a BufferedReader for efficient I/O when handling
     * large input files. Blank lines are skipped.
     * 
     * @return an ArrayList of integers read from stdin
     * @throws IOException if an I/O error occurs while reading
     */
    public static ArrayList<Integer> readInput() throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        ArrayList<Integer> numbers = new ArrayList<>();
        String line;

        while ((line = reader.readLine()) != null) {
            line = line.trim();
            if (!line.isEmpty()) {
                numbers.add(Integer.parseInt(line));
            }
        }

        return numbers;
    }

    /**
     * Sorts a portion of an array in ascending order using the
     * Quick Sort algorithm.
     * 
     * Quick Sort is a divide-and-conquer algorithm that selects a
     * pivot element, partitions the array so that elements less than
     * the pivot come before it and elements greater come after it,
     * then recursively sorts the two partitions. It has O(n log n)
     * average time complexity and O(n^2) worst case.
     * 
     * @param arr  the array of integers to sort
     * @param low  the starting index of the portion to sort (inclusive)
     * @param high the ending index of the portion to sort (inclusive)
     */
    public static void quickSort(int[] arr, int low, int high) {
        if (low < high) {
            int pivotIndex = partition(arr, low, high);

            // Recursively sort elements before and after the pivot
            quickSort(arr, low, pivotIndex - 1);
            quickSort(arr, pivotIndex + 1, high);
        }
    }

    /**
     * Partitions a portion of the array around a pivot element
     * using the Lomuto partition scheme.
     * 
     * The last element in the range is chosen as the pivot. All
     * elements smaller than or equal to the pivot are moved to
     * the left side, and all elements greater are moved to the
     * right side. The pivot is then placed in its correct sorted
     * position.
     * 
     * @param arr  the array to partition
     * @param low  the starting index of the portion to partition
     * @param high the ending index of the portion to partition (pivot)
     * @return the final index of the pivot element
     */
    public static int partition(int[] arr, int low, int high) {
        int pivot = arr[high];
        int i = low - 1;

        for (int j = low; j < high; j++) {
            if (arr[j] <= pivot) {
                i++;

                // Swap arr[i] and arr[j]
                int temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;
            }
        }

        // Place the pivot in its correct position
        int temp = arr[i + 1];
        arr[i + 1] = arr[high];
        arr[high] = temp;

        return i + 1;
    }

    /**
     * Prints each integer in the array to standard output,
     * one per line.
     * 
     * Uses a BufferedWriter for efficient I/O when handling
     * large output.
     * 
     * @param arr the sorted array of integers to print
     * @throws IOException if an I/O error occurs while writing
     */
    public static void printOutput(int[] arr) throws IOException {
        BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(System.out));

        for (int num : arr) {
            writer.write(Integer.toString(num));
            writer.newLine();
        }

        writer.flush();
    }

    /**
     * Main method that orchestrates the Quick Sort program.
     * 
     * Reads integers from standard input, converts them to an array,
     * sorts them using Quick Sort, and prints the sorted result to
     * standard output. This program is designed to be used with
     * input/output redirection, for example:
     * java QuickSort < input.txt > output.txt
     * 
     * @param args command-line arguments (not used)
     */
    public static void main(String[] args) {
        try {
            ArrayList<Integer> numbers = readInput();

            // Convert ArrayList to int array for efficient recursive sorting
            int[] arr = new int[numbers.size()];
            for (int i = 0; i < numbers.size(); i++) {
                arr[i] = numbers.get(i);
            }

            quickSort(arr, 0, arr.length - 1);
            printOutput(arr);
        } catch (IOException e) {
            System.err.println("Error: " + e.getMessage());
            System.exit(1);
        }
    }
}

File: sorting/rust/quick_sort.rs
/*
   Course: CS 41600
   Professor: Dr. Dai
   Date: 2/24/2026
   File: quickSort.rs

   Description:
      This program implements the Quick Sort algorithm to sort a list of
      integers. It reads integers from standard input (one per line), sorts
      them in ascending order using Quick Sort with the Lomuto partition
      scheme, and prints the sorted result to standard output (one integer
      per line). Designed for use with input/output redirection and
      benchmarking.

   Usage:
      ./quickSort < input.txt > output.txt
      cat input.txt | ./quickSort
*/

use std::io::{self, BufRead, BufWriter, Write};
use std::process;

/// Reads integers from standard input, one per line.
///
/// Uses a buffered reader for efficient I/O when handling large input files.
/// Empty lines and whitespace-only lines are skipped. The function will
/// return an error if a line cannot be parsed as an i32.
///
/// # Returns
/// * `io::Result<Vec<i32>>` - A vector of integers read from stdin
///
/// # Errors
/// * Returns an error if reading from stdin fails or if a line cannot be
///   parsed as an integer
fn read_input() -> io::Result<Vec<i32>> {
    let stdin = io::stdin();
    let reader = stdin.lock();
    let mut numbers = Vec::new();

    for line in reader.lines() {
        let line = line?;
        let trimmed = line.trim();
        
        if !trimmed.is_empty() {
            match trimmed.parse::<i32>() {
                Ok(num) => numbers.push(num),
                Err(e) => {
                    eprintln!("Error parsing '{}': {}", trimmed, e);
                    return Err(io::Error::new(
                        io::ErrorKind::InvalidData,
                        format!("Failed to parse integer: {}", trimmed)
                    ));
                }
            }
        }
    }

    Ok(numbers)
}

/// Sorts a slice of integers in ascending order using the Quick Sort algorithm.
///
/// Quick Sort is a divide-and-conquer algorithm that works by:
/// 1. **Select**: Choose a pivot element from the array
/// 2. **Partition**: Rearrange array so elements < pivot come before it,
///    and elements > pivot come after it
/// 3. **Conquer**: Recursively sort the subarrays on either side of the pivot
///
/// This implementation uses the Lomuto partition scheme with the last element
/// as the pivot. The algorithm sorts in-place for O(log n) space complexity
/// (due to recursion stack in the average case).
///
/// # Time Complexity
/// * Best case: O(n log n) when partitions are balanced
/// * Average case: O(n log n)
/// * Worst case: O(n²) when array is already sorted or reverse sorted
///   (with last-element pivot)
///
/// # Space Complexity
/// * O(log n) average case - recursion stack depth
/// * O(n) worst case - recursion stack depth for unbalanced partitions
///
/// # Advantages
/// * In-place sorting (no extra array needed like merge sort)
/// * Cache-friendly due to sequential access patterns
/// * Very fast in practice for random data
/// * Good average-case performance
///
/// # Disadvantages
/// * Not stable (relative order of equal elements may change)
/// * Worst-case O(n²) performance on sorted/reverse-sorted data
/// * Recursive implementation can cause stack overflow on very large arrays
///
/// # Arguments
/// * `numbers` - A mutable slice of i32 integers to sort
pub fn quick_sort(numbers: &mut [i32]) {
    let len = numbers.len();
    if len <= 1 {
        return;
    }
    quick_sort_helper(numbers, 0, (len - 1) as isize);
}

/// Helper function for quick sort that performs the recursive sorting.
///
/// # Arguments
/// * `arr` - The array slice to sort
/// * `low` - Starting index of the portion to sort (inclusive)
/// * `high` - Ending index of the portion to sort (inclusive)
fn quick_sort_helper(arr: &mut [i32], low: isize, high: isize) {
    if low < high {
        // Partition the array and get the pivot index
        let pivot_index = partition(arr, low, high);
        
        // Recursively sort elements before and after the pivot
        quick_sort_helper(arr, low, pivot_index - 1);
        quick_sort_helper(arr, pivot_index + 1, high);
    }
}

/// Partitions a portion of the array around a pivot element using the
/// Lomuto partition scheme.
///
/// The last element in the range is chosen as the pivot. All elements
/// less than or equal to the pivot are moved to the left side, and all
/// elements greater are moved to the right side. The pivot is then
/// placed in its correct sorted position.
///
/// After partitioning:
/// * Elements at indices [low..pivot_index-1] are ≤ pivot
/// * Element at index pivot_index is the pivot (in final position)
/// * Elements at indices [pivot_index+1..high] are > pivot
///
/// # Arguments
/// * `arr` - The array to partition
/// * `low` - Starting index of the portion to partition
/// * `high` - Ending index of the portion to partition (contains pivot)
///
/// # Returns
/// * `isize` - The final index of the pivot element
fn partition(arr: &mut [i32], low: isize, high: isize) -> isize {
    let pivot = arr[high as usize];
    let mut i = low - 1;
    
    // Iterate through the range, moving smaller elements to the left
    for j in low..high {
        if arr[j as usize] <= pivot {
            i += 1;
            // Swap arr[i] and arr[j]
            arr.swap(i as usize, j as usize);
        }
    }
    
    // Place the pivot in its correct position
    arr.swap((i + 1) as usize, high as usize);
    
    i + 1
}

/// Prints each integer in the vector to standard output, one per line.
///
/// Uses a buffered writer for efficient I/O when handling large output.
///
/// # Arguments
/// * `numbers` - A slice of integers to print
///
/// # Returns
/// * `io::Result<()>` - Ok if successful, Err if writing fails
///
/// # Errors
/// * Returns an error if writing to stdout fails
fn print_output(numbers: &[i32]) -> io::Result<()> {
    let stdout = io::stdout();
    let mut writer = BufWriter::new(stdout.lock());

    for &num in numbers {
        writeln!(writer, "{}", num)?;
    }

    writer.flush()?;
    Ok(())
}

/// Main entry point for the quick sort program.
///
/// Orchestrates reading from stdin, sorting, and writing to stdout.
/// Exits with status code 1 if any errors occur during execution.
fn main() {
    // Read integers from standard input
    let mut numbers = match read_input() {
        Ok(nums) => nums,
        Err(e) => {
            eprintln!("Error reading input: {}", e);
            process::exit(1);
        }
    };

    // Sort the numbers using quick sort
    quick_sort(&mut numbers);

    // Print the sorted numbers to standard output
    if let Err(e) = print_output(&numbers) {
        eprintln!("Error writing output: {}", e);
        process::exit(1);
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_empty_array() {
        let mut arr: Vec<i32> = vec![];
        quick_sort(&mut arr);
        assert_eq!(arr, vec![]);
    }

    #[test]
    fn test_single_element() {
        let mut arr = vec![42];
        quick_sort(&mut arr);
        assert_eq!(arr, vec![42]);
    }

    #[test]
    fn test_already_sorted() {
        let mut arr = vec![1, 2, 3, 4, 5];
        quick_sort(&mut arr);
        assert_eq!(arr, vec![1, 2, 3, 4, 5]);
    }

    #[test]
    fn test_reverse_sorted() {
        let mut arr = vec![5, 4, 3, 2, 1];
        quick_sort(&mut arr);
        assert_eq!(arr, vec![1, 2, 3, 4, 5]);
    }

    #[test]
    fn test_random_order() {
        let mut arr = vec![3, 1, 4, 1, 5, 9, 2, 6];
        quick_sort(&mut arr);
        assert_eq!(arr, vec![1, 1, 2, 3, 4, 5, 6, 9]);
    }

    #[test]
    fn test_duplicates() {
        let mut arr = vec![5, 2, 8, 2, 9, 1, 5, 5];
        quick_sort(&mut arr);
        assert_eq!(arr, vec![1, 2, 2, 5, 5, 5, 8, 9]);
    }

    #[test]
    fn test_negative_numbers() {
        let mut arr = vec![-3, 5, -1, 0, -9, 2];
        quick_sort(&mut arr);
        assert_eq!(arr, vec![-9, -3, -1, 0, 2, 5]);
    }

    #[test]
    fn test_large_numbers() {
        let mut arr = vec![1000000, -1000000, 0, 999999, -999999];
        quick_sort(&mut arr);
        assert_eq!(arr, vec![-1000000, -999999, 0, 999999, 1000000]);
    }

    #[test]
    fn test_two_elements_sorted() {
        let mut arr = vec![1, 2];
        quick_sort(&mut arr);
        assert_eq!(arr, vec![1, 2]);
    }

    #[test]
    fn test_two_elements_unsorted() {
        let mut arr = vec![2, 1];
        quick_sort(&mut arr);
        assert_eq!(arr, vec![1, 2]);
    }

    #[test]
    fn test_all_same_elements() {
        let mut arr = vec![7, 7, 7, 7, 7];
        quick_sort(&mut arr);
        assert_eq!(arr, vec![7, 7, 7, 7, 7]);
    }

    #[test]
    fn test_large_array() {
        let mut arr: Vec<i32> = (0..1000).rev().collect();
        quick_sort(&mut arr);
        let expected: Vec<i32> = (0..1000).collect();
        assert_eq!(arr, expected);
    }

    #[test]
    fn test_partition_simple() {
        let mut arr = vec![3, 1, 4, 1, 5];
        let pivot_idx = partition(&mut arr, 0, 4);
        // After partition with pivot=5 (last element):
        // All elements before pivot_idx should be <= 5
        // Element at pivot_idx should be 5
        // All elements after pivot_idx should be > 5
        assert_eq!(arr[pivot_idx as usize], 5);
        for i in 0..pivot_idx as usize {
            assert!(arr[i] <= 5);
        }
    }
}
